### ISSUE TITLE
[Bug] Changing kernel hyperparametes in eval mode results in incorrect predictions

### ISSUE THREAD

**Issue by AustinT** [cornellius-gp/gpytorch#1556]
# 🐛 Bug

Assigning new values to the hyperparameters while a GP is in eval mode seems to lead to incorrect predictions. I believe it is because *`model.prediction_strategy` is not being updated when the kernel hyperparameters are changed*, leading to incorrect cached values. I suggest that to fix this, `model.prediction_strategy` should be reset to None whenever the hyperparameter values are changed.

## To reproduce

First part is a condensed version of the simple GP regression notebook (https://github.com/cornellius-gp/gpytorch/blob/master/examples/01_Exact_GPs/Simple_GP_Regression.ipynb).
```python
import math
import torch
import gpytorch
from matplotlib import pyplot as plt


train_x = torch.linspace(0, 1, 100)
train_y = torch.sin(train_x * (2 * math.pi)) + torch.randn(train_x.size()) * math.sqrt(0.04)

# We will use the simplest form of GP model, exact inference
class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(ExactGPModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel())
    
    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

# initialize likelihood and model
likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = ExactGPModel(train_x, train_y, likelihood)

# Set parameters to proper value
model.covar_module.outputscale = 0.7359
model.covar_module.base_kernel.lengthscale = 0.277
model.likelihood.noise = 0.035

# Get into evaluation (predictive posterior) mode
model.eval()
likelihood.eval()

# Make initial predictions
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    test_x = torch.linspace(0, 1, 51)
    observed_pred = likelihood(model(test_x))
```

### Part demonstrating error

The following snippet changes the kernel lengthscale and plots the resulting predictions.
```python
# Change the hyperparameter and try to make predictions again
model.covar_module.base_kernel.lengthscale = 1e3
# model.train()
model.eval()

with torch.no_grad(), gpytorch.settings.fast_pred_var():
    test_x = torch.linspace(0, 1, 51)
    observed_pred = likelihood(model(test_x))
    
with torch.no_grad():
    # Initialize plot
    f, ax = plt.subplots(1, 1, figsize=(4, 3))

    # Get upper and lower confidence bounds
    lower, upper = observed_pred.confidence_region()
    # Plot training data as black stars
    ax.plot(train_x.numpy(), train_y.numpy(), 'k*')
    # Plot predictive means as blue line
    ax.plot(test_x.numpy(), observed_pred.mean.numpy(), 'b')
    # Shade between the lower and upper confidence bounds
    ax.fill_between(test_x.numpy(), lower.numpy(), upper.numpy(), alpha=0.5)
    ax.set_ylim([-3, 3])
    ax.legend(['Observed Data', 'Mean', 'Confidence'])
```

Output plot: **the variance is essentially zero!**
![image](https://user-images.githubusercontent.com/22754579/114025442-048ae200-986d-11eb-8a01-e19c0b026360.png)

Output plot *if `model.train()` is uncommented* in snippet above: **variance looks correct**

![image](https://user-images.githubusercontent.com/22754579/114025550-23897400-986d-11eb-80b3-73b3268bb2d2.png)

## Expected Behavior

Changing the lengthscale should lead to constant predictions with constant non-zero variance, as shown in the second plot above.

## System information

**Please complete the following information:**
-  1.4.0
-  1.8.1
-  Ubuntu 18.04

**Comment by gpleiss** [comment]
We compute caches to speed up predictions in `eval` mode. When you change the lengthscale, the model is still using caches from the old lengthscale value.

To update the caches, one hack is to call `model.train(); # code to set lengthscale value; model.eval()`. Going into training mode clears all caches.

**Comment by jacobrgardner** [comment]
@AustinT I think clearing the caches when the setters are called makes sense. An even simpler fix would be to have the setters throw an error if the module is in `eval` mode, telling the user to switch to `train` mode before using them.

We already try to handle this to some extent -- if the parameters get updated by backpropagation, they get cleared: https://github.com/cornellius-gp/gpytorch/blob/405f0bdafd2d0767dc736f9d27350538f0ac635d/gpytorch/models/exact_prediction_strategies.py#L82-L85

**Comment by gpleiss** [comment]
@AustinT a PR would be 100% appreciated!
